core/trino-main/src/main/java/io/trino/operator/InterpretedHashGenerator.java

```diff
@@ -45,26 +48,45 @@ public InterpretedHashGenerator(List<Type> hashChannelTypes, List<Integer> hashC
 
     public InterpretedHashGenerator(List<Type> hashChannelTypes, int[] hashChannels, BlockTypeOperators blockTypeOperators)
     {
-        this.hashChannels = requireNonNull(hashChannels, "hashChannels is null");
+        this(hashChannelTypes, requireNonNull(hashChannels, "hashChannels is null"), blockTypeOperators, false);
+    }
+
+    private InterpretedHashGenerator(List<Type> hashChannelTypes, @Nullable int[] hashChannels, BlockTypeOperators blockTypeOperators, boolean positional)
+    {
         this.hashChannelTypes = ImmutableList.copyOf(requireNonNull(hashChannelTypes, "hashChannelTypes is null"));
-        checkArgument(hashChannelTypes.size() == hashChannels.length);
-        this.hashCodeOperators = hashChannelTypes.stream()
-                .map(blockTypeOperators::getHashCodeOperator)
-                .collect(toImmutableList());
+        this.hashCodeOperators = createHashCodeOperators(hashChannelTypes, blockTypeOperators);
+        checkArgument(hashCodeOperators.length == hashChannelTypes.size());
+        if (positional) {
+            checkArgument(hashChannels == null, "hashChannels must be null");
+            this.hashChannels = null;
+        }
+        else {
+            requireNonNull(hashChannels, "hashChannels is null");
+            checkArgument(hashChannels.length == hashCodeOperators.length);
+            // simple positional indices are converted to null
+            this.hashChannels = isPositionalChannels(hashChannels) ? null : hashChannels;
+        }
     }
 
     @Override
     public long hashPosition(int position, Page page)
     {
-        return hashPosition(position, page::getBlock);
+        // Note: this code is duplicated for performance but must logically match hashPosition(position, IntFunction<Block> blockProvider)
```

Comment:
does splitting methods make sense? There are only two implementations of `IntFunction<Block> blockProvider`, so it should not be megamorphic. Anyway, obtaining `Block` doesn't seem that ciritical

Reply:
Unfortunately it does seem to require that (at least on JDK8). The loop unswitching transformation is very sensitive to whether the inputs to the loop might "escape" and putting an `IntFunction<Block>` with more than 1 implementation seems to make it pessimistic about the possibility that some reflection code somewhere might be altering the loop condition (eg: via reflection to overwrite `hashChannels`). Honestly, it might be better to change the other method to just take a `PageBuilder` instead of using `IntFunction` since I think that's the only other argument type and at least that would be concrete.

Reply:
I don't think we care about JDK8 anymore. 11 is min